develop/2016-12-30-fgrehm-ocr-receipts-with-google-cloud-vision.py

```diff
@@ -0,0 +1,168 @@
+# 
+# On the Docker environment provided, you'll need to `docker exec -u root -ti CONTAINER bash` in order to have permissions to install the packages with a
+# 
+# ```
+# apt-get update && apt-get install -y poppler-utils jq
+# ```
+# 
+# ## Download receipts, convert to PNG and OCR them
+
+# In[2]:
+
+get_ipython().run_cell_magic('bash', '', '\nocr() {\n  id="${1}"\n  url="${2}"\n  mkdir -p "/tmp/reimbursements/${id}"\n  cd "/tmp/reimbursements/${id}"\n  echo "---> $id"\n  echo "     Downloading PDF from \'$url\'..."\n  curl -s "${url}" > "document.pdf"\n  echo "     Generating PNGs..."\n  pdftoppm -rx 300 -ry 300 -png "document.pdf" page\n  \n  for img in page*.png; do\n    echo "     OCRing ${img}..."\n    payload="payload-${img%.*}.json"\n    response="response-${img%.*}.json"\n    echo -n \'{"requests": [ { "features": [ { "type": "TEXT_DETECTION" } ], "image": { "content": "\' > $payload\n    base64 -w 0 $img >> $payload\n    echo -n \'" } } ] }\' >> $payload\n    \n    curl -s "https://vision.clients6.google.com/v1/images:annotate?key=$(cat /tmp/cloud-vision.key)&alt=json" \\\n         --data-binary @$payload \\\n         -H \'Content-Type: application/json\' \\\n      > $response\n  done\n}\n\ndate\nocr 5631309 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/1789/2015/5631309.pdf\'\nocr 5631380 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/1789/2015/5631380.pdf\'\nocr 5928875 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/1564/2016/5928875.pdf\'\nocr 5768932 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/80/2015/5768932.pdf\'\nocr 5962849 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/3052/2016/5962849.pdf\'\nocr 5962903 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/3052/2016/5962903.pdf\'\nocr 5855221 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/2238/2015/5855221.pdf\'\nocr 5856784 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/2238/2015/5856784.pdf\'\nocr 5921187 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/2871/2016/5921187.pdf\'\nocr 6069360 \'http://www.camara.gov.br/cota-parlamentar/documentos/publ/2935/2016/6069360.pdf\'\ndate')
+
+
+# As we can see, it takes a while to process just 10 PDFs on a 60Mb connection (5 minutes), if we ever move on with this we should really look into parallelizing it from day 0 and / or sending receipts in batches as it is supported by the API.
+# 
+# ## Document [5631309](https://jarbas.datasciencebr.com/#/document_id/5631309)
+# 
+# A hotel receipt with lots of text in it. There is a lot of stuff on those JSON responses so we just extract the info for the piece that represents the whole text.
+
+# In[3]:
+
+get_ipython().run_cell_magic('bash', '', "jq -r '.responses[].textAnnotations[0].description' /tmp/reimbursements/5631309/response-page-1.json\necho '--------------'; echo\njq -r '.responses[].textAnnotations[0].description' /tmp/reimbursements/5631309/response-page-2.json")
```

Comment:
For comparison, using the same method as https://github.com/datasciencebr/serenata-de-amor/commit/fb93f96e334c46f98eea3d4a9db565b8bc6bb45b

```python
import os.path
import subprocess

from PIL import Image
from pytesseract import image_to_string

def pdf_to_image(input_path):
  output_path = input_path.replace('pdf', 'png')
  if not os.path.exists(output_path):
    subprocess.call(['sips', '-s', 'format', 'png',
                     input_path, '--out', output_path])

pdf_to_image('5631309.pdf')
image_to_string(Image.open(input_path), config='-psm 6')
```

gives the following output:

```
Castro Marques Hotels LTDA.
Extrato de Conta
Nome: omvo BILAC 911410 Uh: 1001
Empress: 511311an Ranerva; 1820028
mm can: mass. Iiscal : micrl/CzZ: mom
EnderegomuA 19151115110 mums mum, 4n Apto sol . savanna FunaionéxiozADSIL‘lA
mm HORIZONTE NC 30320-5911 3511st 211113350: 13/03/2015 no t 35
Chagada: 12/c'3/2m: 00:22 Pam’s; 14/03/2915 new:
1
N
CONTA ENCEHHADA Héspeae: PINTO‘ OLAVO BHAC Num Dom «55mm DnslgnagaoﬁLAVO BILAC PINTO
Olin 07mm Documemo on Mamas) Empress Seldo Usuéria
:s/msm m: Iguanas-vacuum mums 71,00 vs,» um:
mm m mm may»: mun noon a“ man ~ awn
Damn/w: 1m - 11,00
momma can. mum “mus "Immun- nma um: ALuzz
Jam am on "cam “1‘59 - 25,17: 1 .
Am cm was 2mm!» :a 4,0: - 12,3:
H/OJI‘IZ? COOI ”VHSNIIII'AWXI‘TI IUWB 259100 381700 NJ!!!
am: warn ms 341m mm In A 1;: 15,170 - 115, cc
:15me as Maw Mam ’Wm H M95 - ma:-
"mm mug mm: mm mm as Inna: - 4m:
(Gan REM ms new?! Jar/LN - 9,09
Aom cm at}; 3:55:12; 2; 4,69 - 9,5:
sacs: 1:; may; {4'91 545: = 1m:
1 as: name by 3,3: - 1mg
1 ”names com 63021119924 921212112; memmw 34,417 nun uwzz
‘ saw mu m 310L513 1x345 :- 14, :0
‘ 3C0”). (UK £1.43 RIOLEVX 5x @530 B 26%;};
:zwsmsa ego: ssuzsrxmmanz 1,0; 123,113 "shunt!
KEY/A C05 “3 BICLEW l x {.06 ~ «351?
13/032365 0001 6903a “kam'tﬂ wanna 192ml} tit/(.00 NNRGEH
5030 as 45mm 5»: 5,50 = 27.37
311m vb maven 3205,50 a 15:55
SizesSrM mt we max/mo zxsmw = 13,051
prom Am- max new saga; a; 1x45, 0: » 49,00
awn G952! a0 mm; can any cs lx§5,&'9 - 45.»
can: mam: a); 3,259 .. no:
1419305163 cm: 69638 X008 SERVICE RESTAUHKRT 50.9%} 19¢,” nut:
Elf-K P1. KLVM MIRCIJA'E PLAZA (6634’ 3X 30190 = 96070
11/03 09:03 xxwxmxxxx. DEPOSIN ANTECIFMEO ~165,€0 ~‘!.IB 1|le
[claim 439133
Resume do Extrato
Héspede: PINTO. OLAVO an Designaoﬁo: OLAVO BHAC PWTO
Descrigéo Valor Total
DEPOSITO AWECIFA 46:. do
FEIGDEAR ;,oa
RES7AUMNTE UJGAN 52a, 00
ROOM same: REST 36¢, 09
R.—-\___U._un_. . K
```